src/integrations/pulsar/consumer.cpp

```diff
@@ -277,11 +275,66 @@ void Consumer::StartConsuming() {
   });
 }
 
+void Consumer::StartConsumingWithLimit(int64_t limit_batches) const {
+  if (is_running_.exchange(true)) {
+    throw ConsumerRunningException(info_.consumer_name);
+  }
+  utils::OnScopeExit restore_is_running([this] { is_running_.store(false); });
+
+  for (int64_t i = 0; i < limit_batches;) {
+    auto maybe_batch = GetBatch(consumer_, info_, is_running_, last_message_id_);
+
+    if (maybe_batch.HasError()) {
+      spdlog::warn("Error happened in consumer {} while fetching messages: {}!", info_.consumer_name,
+                   maybe_batch.GetError());
+      break;
+    }
+
+    const auto &batch = maybe_batch.GetValue();
+
+    if (batch.empty()) {
+      continue;
+    }
+    ++i;
+
+    spdlog::info("Pulsar consumer {} is processing a batch", info_.consumer_name);
+
+    const auto consumed = TryToConsumeBatch(batch);
+    if (!consumed) {
+      return;
+    }
+
+    spdlog::info("Pulsar consumer {} finished processing", info_.consumer_name);
+  }
+}
+
 void Consumer::StopConsuming() {
   is_running_.store(false);
   if (thread_.joinable()) {
     thread_.join();
   }
 }
 
+bool Consumer::TryToConsumeBatch(const std::vector<Message> &batch) const {
```

Comment:
I would use exception based error handling here also because of the same reason as in the Kafka producer (it propagates to the users in the blocking calls).